### PR TITLE
RS encryption and signature tests

### PR DESCRIPTION
**Pull request by c00crane**
Add tests to validate RS handling of JWS' using newly supported signature algorithms and JWE's.
Create JWTs with the JWT Builder or test code and use those to access a test app protected by openidconnect client acting as an RS (not an RP).
The JWTs created will use newly supported signature algorithms - these tests will use the following signature algorithms:
HS256, HS384, HS512
RS256, RS384, RS512
ES256, ES384, ES512
Valid and invalid combinations of signing and consuming are tested.

JWTs created will use the following algorithms for encryption:
RS256, RS384, RS512
ES256, ES384, ES512
Valid and invalid combinations of encryption and decryption are tested.  Tests that sign and encrypt in all combinations are also included.

### CODE REVIEW

**Comment by c00crane** [comment]
#run-libby-bot
#build

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_sigAlg.xml [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+	<include location="${server.config.dir}/imports/BasicRegistry.xml" />
+	<include location="${server.config.dir}/imports/goodSSLSettings.xml" />
+	<include location="${server.config.dir}/imports/OPmisc.xml" />
+	<include location="${server.config.dir}/imports/jwtBuilderApplication.xml" />
+
+	<featureManager>
+		<feature>oauth-2.0</feature>
+		<feature>ssl-1.0</feature>
+		<feature>securitylibertyinternals-1.0</feature>
+		<feature>openidConnectServer-1.0</feature>
```
Comment (ayoho) [inline dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_sigAlg.xml]: We shouldn't need the OIDC feature, right?

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_sigAlg.xml [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+		/>
+		
+	<jwtBuilder 
+		id="RS256Builder"
+		issuer="client01"
+		signatureAlgorithm="RS256"
+		audiences="client01"
+		keyStoreRef="key_allSigAlg"
+        keyAlias="rs256" />
+		/>
```
Comment (ayoho) [inline dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_sigAlg.xml]: You've got some invalid XML here and in the rest of the builders.

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+			urlPattern="/snoop/HS256"
+			matchType="contains" />
+	</authFilter>
+
+	<openidConnectClient
+		id="hs256"
+		scope="openid profile"
+		clientId="client01"
+		clientSecret="{xor}LDo8LTor"
+		sharedKey="secret"
```
Comment (ayoho) [inline dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml:38]: Not needed

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+		mapIdentityToRegistryUser="true"
+		httpsRequired="false"
+		redirectToRPHostAndPort="http://localhost:${bvt.prop.security_3_HTTP_default}"
+		validateAccessTokenLocally="true"
```
Comment (ayoho) [inline dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml]: Not needed

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml [hunk 5]
```diff
@@ -0,0 +1,1838 @@
...
+		keyManagementKeyAlias="rs256"
+		sslRef="ssl_allSigAlg"
+		inboundPropagation="required"
+		issuerIdentifier="client01"
+		audiences="client01"
+	>
+	</openidConnectClient>	
+
+	<openidConnectClient
+		id="SignRS256EncryptRS256"
```
Comment (ayoho) [inline dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml]: You've duplicated this config

**Comment by c00crane** [comment]
#run-libby-bot
#build

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml [hunk 6]
```diff
@@ -0,0 +1,1838 @@
...
+
+	<authFilter id="authFilterRS256none">
+		<requestUrl
+			id="requestUrlRS256none"
+			urlPattern="/snoop/SignRS256Encryptnone"
+			matchType="contains" />
+	</authFilter>
+
+	<openidConnectClient
+		id="SignRS256Encryptnone"
```
Comment (ayoho) [inline dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml:1565]: Are the `Encryptnone` configs/tests any different than the configs tested by the signature algorithm tests?

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml [hunk 7]
```diff
@@ -0,0 +1,749 @@
...
+	<include location="${server.config.dir}/imports/BasicRegistry.xml" />
+	<include location="${server.config.dir}/imports/goodSSLSettings.xml" />
+	<include location="${server.config.dir}/imports/OPmisc.xml" />
+	<include location="${server.config.dir}/imports/jwtBuilderApplication.xml" />
+
+	<featureManager>
+		<feature>oauth-2.0</feature>
+		<feature>ssl-1.0</feature>
+		<feature>securitylibertyinternals-1.0</feature>
+		<feature>openidConnectServer-1.0</feature>
```
Comment (ayoho) [inline dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml]: Is the OIDC feature needed?

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml [hunk 8]
```diff
@@ -0,0 +1,749 @@
...
+<!-- ES Encryption builders -->	
+	<jwtBuilder 
+		id="SignRS256EncryptES256Builder"
+		issuer="client01"
+		signatureAlgorithm="RS256"
+		audiences="client01"
+		keyStoreRef="key_allSigAlg"
+        keyAlias="rs256"
+        trustStoreRef="trust_allSigAlg"
+		keyManagementKeyAlgorithm="ECDH-ES"
```
Comment (ayoho) [inline dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml:372]: I'm reviewing the XMLs before the tests, so I'm sure I'll see later whether you do have this addressed: This option isn't supported yet, so you'll get a warning message in the server during startup for all of these configs that we'll have to account for.

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/fat/src/com/ibm/ws/security/openidconnect/server/fat/jaxrs/config/noOP/NoOPSignatureRSServerTests.java [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+        // set token and cert types - hard code JWT and then select the token type to use...
+        String tokenType = Constants.JWT_TOKEN;
+        //		String certType = rsTools.chooseCertType(tokenType) ;
+        String certType = Constants.X509_CERT;
+
+        testSettings = new TestSettings();
+
+        testOPServer = commonSetUp(OPServerName, "server_sigAlg.xml", Constants.OIDC_OP, extraApps, Constants.DO_NOT_USE_DERBY, extraMsgs, null, Constants.OIDC_OP, true, true, tokenType, certType);
+        SecurityFatHttpUtils.saveServerPorts(testOPServer.getServer(), Constants.BVT_SERVER_1_PORT_NAME_ROOT);
+        testOPServer.addIgnoredServerException(MessageConstants.CWWKG0032W_CONFIG_INVALID_VALUE + ".*" + "ECDH-ES");
```
Comment (ayoho) [inline dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/fat/src/com/ibm/ws/security/openidconnect/server/fat/jaxrs/config/noOP/NoOPSignatureRSServerTests.java:81]: 👍

**Review by ayoho — APPROVED** [review]

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml (hunk 6 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml:1565]: yes - they're used for encrypted tokens being sent to a client that do not have encryption configureud.

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml (hunk 7 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml]: Could probably replace oidc and oauth with jwt (which is needed for the builder)
I can do that via another pull

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml (hunk 8 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml:372]: I've added a comment to the issue adding support for ECDH-ES to remove all of the work arounds (setting that value via the api's and allowing the warning for the unrecognized value...

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml (hunk 5 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml]: oh how could that ever happen (with so many similar configs :) )

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_sigAlg.xml (hunk 1 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_sigAlg.xml]: Could probably replace oidc and oauth with jwt (which is needed for the builder)
I can do that via another pull

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml (hunk 3 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml:38]: checking what can be removed (to test with the minimum config)

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml (hunk 4 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml]: checking what can be removed (to test with the minimum config)

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml (hunk 7 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_encryption.xml]: Replaced with:
	<featureManager>
		<feature>ssl-1.0</feature>
		<feature>jwt-1.0</feature>
		<feature>jsonp-1.0</feature>
	</featureManager>

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_sigAlg.xml (hunk 1 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.OPserver/configs/server_sigAlg.xml]: Replaced with:
	<featureManager>
		<feature>ssl-1.0</feature>
		<feature>jwt-1.0</feature>
		<feature>jsonp-1.0</feature>
	</featureManager>

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml (hunk 3 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml:38]: need the shared key (in the HS configs), but not the other settings.

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml (hunk 4 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_sigAlg.xml]: These can be removed

File: dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml (hunk 5 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.oidc.server_fat.jaxrs.config/publish/servers/com.ibm.ws.security.openidconnect.server-1.0_fat.jaxrs.config.RSserver/configs/server_encryption.xml]: dup removed

**Comment by c00crane** [comment]
#run-libby-bot
#build

**Pull request merged**